File: Second year/Design and Programming Environments/TicketManagerCSharp/TicketManagerCSharp/repository/UserRepository.cs
using log4net;
using System;
using System.Collections.Generic;
using System.Data;
using TicketManagerCSharp.domain;
using TicketManagerCSharp.utils;

namespace TicketManagerCSharp.repository
{
    public class UserRepository : IRepository<string, User>
    {
        private static readonly ILog log = LogManager.GetLogger("EventRepository");

        private IDictionary<String, string> props;

        public UserRepository(IDictionary<String, string> props)
        {
            log.Info("Creating UserRepository...");
            this.props = props;
        }

        public void Delete(string id)
        {
            log.InfoFormat("Entering Delete with new value {0}...", id);
            IDbConnection con = DBUtils.getConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "delete from Users where username=@id";
                IDbDataParameter idParam = comm.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = id;
                comm.Parameters.Add(idParam);
                var rez = comm.ExecuteNonQuery();
                if (rez == 0)
                {
                    throw new RepositoryException("Error:Can't delete User!");
                }
                log.InfoFormat("Exiting Delete");
            }
        }

        public IEnumerable<User> FindAll()
        {
            IDbConnection con = DBUtils.getConnection(props);
            log.InfoFormat("Entering findAll...");
            IList<User> users = new List<User>();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select username,password from Users";
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        string username = dataR.GetString(0);
                        string password = dataR.GetString(1);
                        User user = new User(username, password);
                        users.Add(user);
                    }
                }
            }
            log.InfoFormat("Exiting findAll");
            return users;
        }

        public User FindOne(string username)
        {
            log.InfoFormat("Entering findOne with value {0}", username);
            IDbConnection con = DBUtils.getConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select username,password from Users where username=@username";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@username";
                paramId.Value = username;
                comm.Parameters.Add(paramId);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        string newUsername = dataR.GetString(0);
                        string password = dataR.GetString(1);
                        User user = new User(newUsername, password);
                        log.InfoFormat("Exiting findOne with value{0}", user);
                        return user;
                    }
                }
            }
            log.InfoFormat("Exiting findOne with value {0}", null);
            return null;
        }

        public void Save(User entity)
        {
            log.InfoFormat("Entering Save with new value {0}...", entity);
            var con = DBUtils.getConnection(props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "insert into Users(username,password) values (@username,@password)";
                var usernameParam = comm.CreateParameter();
                usernameParam.ParameterName = "@username";
                usernameParam.Value = entity.Id;
                comm.Parameters.Add(usernameParam);

                var passwordParam = comm.CreateParameter();
                passwordParam.ParameterName = "@password";
                passwordParam.Value = entity.Password;
                comm.Parameters.Add(passwordParam);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    throw new RepositoryException("Error:Can't add user!");
                }
                log.InfoFormat("Exiting Save");
            }
        }

        public int Size()
        {
            log.InfoFormat("Calculating Users database Size...");
            IDbConnection con = DBUtils.getConnection(props);
            int size = -1;
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select count(*) from Users";

                var rez = comm.ExecuteNonQuery();
                if (rez == 0)
                {
                    throw new RepositoryException("Error:Can't calculate size!");
                }

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        size = dataR.GetInt32(0);
                    }
                }
            }
            log.InfoFormat("Finished calculation!");
            return size;
        }

        public void Update(string id, User entity)
        {
            log.InfoFormat("Entering Update with value {0}", id);
            var con = DBUtils.getConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "update Userss set password=@password where username=@id";
                var idParam = comm.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = entity.Id;
                comm.Parameters.Add(idParam);

                var passwordParam = comm.CreateParameter();
                passwordParam.ParameterName = "@password";
                passwordParam.Value = entity.Password;
                comm.Parameters.Add(passwordParam);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    throw new RepositoryException("Error:Can't update User!");
                }
                log.InfoFormat("Exiting Update");
            }
        }
    }
}

File: Second year/Design and Programming Environments/TicketManagerJava/Persistance/src/main/java/Repository/UserRepository.java
package Repository;

import ModelDomain.User;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class UserRepository implements IRepository<String,User> {
    private Connection connection;

    public UserRepository() {
        connection = DatabaseConnection.getConnection();
    }

    @Override
    public int size() {
        try(PreparedStatement preStmt=connection.prepareStatement("select count(*) as [SIZE] from Users")){
            try(ResultSet result = preStmt.executeQuery()){
                if(result.next()){
                    return result.getInt("SIZE");
                }
            }
        } catch (SQLException e) {
            System.out.println("Error DB ModelDomain.User Size " + e);
        }
        return 0;
    }

    @Override
    public void save(User entity) {
        try(PreparedStatement preStmt=connection.prepareStatement("insert into Users values (?,?)")) {
            preStmt.setString(1,entity.getId());
            preStmt.setString(2,entity.getPassword());
            int result = preStmt.executeUpdate();
        }
        catch (SQLException e) {
            System.out.println("Error DB ModelDomain.User Save "+e);
        }
    }

    @Override
    public void delete(String s) {
        try(PreparedStatement preStmt=connection.prepareStatement("delete from Users where username=?")){
            preStmt.setString(1,s);
            int result = preStmt.executeUpdate();
        }catch(SQLException e){
            System.out.println("Error DB ModelDomain.User Delete "+e);
        }
    }

    @Override
    public void update(String s, User entity) {
        try(PreparedStatement preStmt=connection.prepareStatement("update Users set password=? where username=?")){
            preStmt.setString(1,entity.getPassword());
            preStmt.setString(2,entity.getId());
            int result=preStmt.executeUpdate();
        }catch (SQLException ex){
            System.out.println("Error DB "+ ex);
        }
    }

    @Override
    public User findOne(String s) {
        try(PreparedStatement preStmt=connection.prepareStatement("select * from Users where username=?")){
            preStmt.setString(1,s);
            try(ResultSet result = preStmt.executeQuery()){
                if(result.next()){
                    String username=result.getString("username");
                    String password=result.getString("password");
                    User user = new User(username,password);
                    return user;
                }
            }
        }catch(SQLException e) {
            System.out.println("Error DB ModelDomain.User FindOne " + e);
        }
        return null;
    }


    @Override
    public Iterable<User> findAll() {
        List<User> users=new ArrayList<>();
        try(PreparedStatement preStmt=connection.prepareStatement("select * from Users")){
            try(ResultSet result=preStmt.executeQuery()){
                while(result.next()){
                    String username=result.getString("username");
                    String password=result.getString("password");
                    User user = new User(username,password);
                    users.add(user);
                }
            }
        }catch(SQLException e) {
            System.out.println("Error DB ModelDomain.User FindAll " + e);
        }
        return users;
    }
}
